File: tools-multiplatform/nodejs/web-scraper.js
#!/usr/bin/env node
const axios = require('axios');
const cheerio = require('cheerio');
const { URL } = require('url');
const { program } = require('commander');
const chalk = require('chalk');

class WebScraper {
    constructor(url) {
        this.url = url;
        this.baseUrl = new URL(url);
        this.results = {
            url: url,
            emails: [],
            phones: [],
            socialLinks: [],
            metadata: {},
            links: { internal: [], external: [] }
        };
    }

    async scrape() {
        console.log(chalk.cyan(`\n[*] Scraping: ${this.url}\n`));
        
        try {
            const response = await axios.get(this.url, {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
                },
                timeout: 10000
            });

            const $ = cheerio.load(response.data);
            
            // Extract various data
            this.extractEmails(response.data);
            this.extractPhones(response.data);
            this.extractMetadata($);
            this.extractLinks($);
            this.extractSocialLinks($);
            
            // Display results
            this.displayResults();
            
            return this.results;
            
        } catch (error) {
            console.log(chalk.red(`[-] Error: ${error.message}`));
            return null;
        }
    }

    extractEmails(html) {
        const emailRegex = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g;
        const emails = [...new Set(html.match(emailRegex) || [])];
        this.results.emails = emails.filter(email => 
            !email.endsWith('.png') && !email.endsWith('.jpg')
        );
    }

    extractPhones(html) {
        const phoneRegexes = [
            /\+?1?\s*\(?[0-9]{3}\)?[\s.-]?[0-9]{3}[\s.-]?[0-9]{4}/g,
            /\+[0-9]{1,3}\s?[0-9]{1,14}/g
        ];
        
        const phones = new Set();
        phoneRegexes.forEach(regex => {
            const matches = html.match(regex) || [];
            matches.forEach(phone => phones.add(phone.trim()));
        });
        
        this.results.phones = [...phones];
    }

    extractMetadata($) {
        // Title
        this.results.metadata.title = $('title').text().trim();
        
        // Meta tags
        $('meta').each((i, elem) => {
            const name = $(elem).attr('name') || $(elem).attr('property');
            const content = $(elem).attr('content');
            if (name && content) {
                this.results.metadata[name] = content;
            }
        });
        
        // Open Graph data
        const ogData = {};
        $('meta[property^="og:"]').each((i, elem) => {
            const property = $(elem).attr('property').replace('og:', '');
            ogData[property] = $(elem).attr('content');
        });
        if (Object.keys(ogData).length > 0) {
            this.results.metadata.openGraph = ogData;
        }
    }

    extractLinks($) {
        $('a[href]').each((i, elem) => {
            const href = $(elem).attr('href');
            if (!href) return;
            
            try {
                const linkUrl = new URL(href, this.baseUrl);
                
                if (linkUrl.hostname === this.baseUrl.hostname) {
                    this.results.links.internal.push(linkUrl.href);
                } else if (linkUrl.protocol.startsWith('http')) {
                    this.results.links.external.push(linkUrl.href);
                }
            } catch (e) {
                // Invalid URL
            }
        });
        
        // Remove duplicates and limit
        this.results.links.internal = [...new Set(this.results.links.internal)].slice(0, 20);
        this.results.links.external = [...new Set(this.results.links.external)].slice(0, 20);
    }

    extractSocialLinks($) {
        const socialPlatforms = {
            'facebook.com': 'Facebook',
            'twitter.com': 'Twitter',
            'linkedin.com': 'LinkedIn',
            'instagram.com': 'Instagram',
            'youtube.com': 'YouTube',
            'github.com': 'GitHub',
            'tiktok.com': 'TikTok'
        };
        
        const socialLinks = new Map();
        
        $('a[href]').each((i, elem) => {
            const href = $(elem).attr('href');
            if (!href) return;
            
            for (const [domain, platform] of Object.entries(socialPlatforms)) {
                if (href.includes(domain)) {
                    socialLinks.set(platform, href);
                }
            }
        });
        
        this.results.socialLinks = Array.from(socialLinks).map(([platform, url]) => ({
            platform, url
        }));
    }

    displayResults() {
        console.log(chalk.green('[+] Scraping Results:'));
        console.log(chalk.yellow('\nMetadata:'));
        console.log(`  Title: ${this.results.metadata.title || 'N/A'}`);
        
        if (this.results.emails.length > 0) {
            console.log(chalk.yellow('\nEmails Found:'));
            this.results.emails.forEach(email => console.log(`  - ${email}`));
        }
        
        if (this.results.phones.length > 0) {
            console.log(chalk.yellow('\nPhone Numbers Found:'));
            this.results.phones.forEach(phone => console.log(`  - ${phone}`));
        }
        
        if (this.results.socialLinks.length > 0) {
            console.log(chalk.yellow('\nSocial Media Links:'));
            this.results.socialLinks.forEach(link => 
                console.log(`  - ${link.platform}: ${link.url}`)
            );
        }
        
        console.log(chalk.yellow('\nLinks Summary:'));
        console.log(`  Internal: ${this.results.links.internal.length} found`);
        console.log(`  External: ${this.results.links.external.length} found`);
    }
}

// CLI setup
program
    .name('web-scraper')
    .description('Web Scraping Tool - Lackadaisical Security')
    .version('1.0.0')
    .argument('<url>', 'URL to scrape')
    .option('-o, --output <file>', 'Save results to JSON file')
    .action(async (url, options) => {
        const scraper = new WebScraper(url);
        const results = await scraper.scrape();
        
        if (results && options.output) {
            const fs = require('fs').promises;
            await fs.writeFile(options.output, JSON.stringify(results, null, 2));
            console.log(chalk.green(`\n[+] Results saved to ${options.output}`));
        }
    });

program.parse();

File: tools/web_scraper.py
import requests
from bs4 import BeautifulSoup
from typing import Dict, List, Any
from urllib.parse import urljoin, urlparse
import re
from config import Config

class WebScraper:
    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update({'User-Agent': Config.USER_AGENT})
        
    def scrape_intel(self, url: str) -> Dict[str, Any]:
        """Scrape intelligence from a website"""
        results = {
            'url': url,
            'emails': [],
            'phone_numbers': [],
            'social_links': [],
            'technologies': [],
            'meta_info': {},
            'links': {
                'internal': [],
                'external': []
            }
        }
        
        try:
            response = self.session.get(url, timeout=Config.REQUEST_TIMEOUT)
            soup = BeautifulSoup(response.text, 'html.parser')
            
            # Extract emails
            results['emails'] = self._extract_emails(response.text)
            
            # Extract phone numbers
            results['phone_numbers'] = self._extract_phones(response.text)
            
            # Extract social media links
            results['social_links'] = self._extract_social_links(soup, url)
            
            # Extract meta information
            results['meta_info'] = self._extract_meta_info(soup)
            
            # Extract and categorize links
            results['links'] = self._extract_links(soup, url)
            
            # Detect technologies
            results['technologies'] = self._detect_technologies(response)
            
        except Exception as e:
            results['error'] = str(e)
            
        return results
    
    def _extract_emails(self, text: str) -> List[str]:
        """Extract email addresses from text"""
        email_pattern = r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}'
        emails = list(set(re.findall(email_pattern, text)))
        return [email for email in emails if not email.endswith('.png') and not email.endswith('.jpg')]
    
    def _extract_phones(self, text: str) -> List[str]:
        """Extract phone numbers from text"""
        phone_patterns = [
            r'\+?1?\s*\(?[0-9]{3}\)?[\s.-]?[0-9]{3}[\s.-]?[0-9]{4}',
            r'\+[0-9]{1,3}\s?[0-9]{1,14}'
        ]
        phones = []
        for pattern in phone_patterns:
            phones.extend(re.findall(pattern, text))
        return list(set(phones))
    
    def _extract_social_links(self, soup: BeautifulSoup, base_url: str) -> List[Dict[str, str]]:
        """Extract social media links"""
        social_platforms = {
            'facebook.com': 'Facebook',
            'twitter.com': 'Twitter',
            'linkedin.com': 'LinkedIn',
            'instagram.com': 'Instagram',
            'youtube.com': 'YouTube',
            'github.com': 'GitHub'
        }
        
        social_links = []
        for link in soup.find_all('a', href=True):
            href = link['href']
            for platform_url, platform_name in social_platforms.items():
                if platform_url in href:
                    social_links.append({
                        'platform': platform_name,
                        'url': urljoin(base_url, href)
                    })
                    
        return social_links
    
    def _extract_meta_info(self, soup: BeautifulSoup) -> Dict[str, str]:
        """Extract meta information"""
        meta_info = {}
        
        # Title
        if soup.title:
            meta_info['title'] = soup.title.string
            
        # Meta tags
        for meta in soup.find_all('meta'):
            if meta.get('name'):
                meta_info[meta['name']] = meta.get('content', '')
            elif meta.get('property'):
                meta_info[meta['property']] = meta.get('content', '')
                
        return meta_info
    
    def _extract_links(self, soup: BeautifulSoup, base_url: str) -> Dict[str, List[str]]:
        """Extract and categorize links"""
        internal_links = []
        external_links = []
        base_domain = urlparse(base_url).netloc
        
        for link in soup.find_all('a', href=True):
            href = urljoin(base_url, link['href'])
            if urlparse(href).netloc == base_domain:
                internal_links.append(href)
            else:
                external_links.append(href)
                
        return {
            'internal': list(set(internal_links))[:20],  # Limit to 20
            'external': list(set(external_links))[:20]
        }
    
    def _detect_technologies(self, response: requests.Response) -> List[str]:
        """Detect technologies used by the website"""
        technologies = []
        
        # Check headers
        if 'X-Powered-By' in response.headers:
            technologies.append(response.headers['X-Powered-By'])
        if 'Server' in response.headers:
            technologies.append(response.headers['Server'])
            
        # Check for common frameworks in HTML
        html_checks = {
            'WordPress': ['wp-content', 'wp-includes'],
            'React': ['react', 'reactdom'],
            'Angular': ['ng-app', 'angular'],
            'Vue.js': ['vue.js', 'vue.min.js'],
            'jQuery': ['jquery.min.js', 'jquery.js']
        }
        
        for tech, patterns in html_checks.items():
            for pattern in patterns:
                if pattern in response.text.lower():
                    technologies.append(tech)
                    break
                    
        return list(set(technologies))
